src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapPlot.jsx

```diff
@@ -58,91 +59,184 @@ const HeatmapPlot = (props) => {
 
     const data = createVegaData(selectedGenes, expressionData);
     setDataDebounce(data);
-  }, [loadingGenes, hidden]);
+  }, [loadingGenes, hidden, properties, hierarchy]);
 
-  const createVegaData = (selected, expression) => {
-    const data = { cellOrder: [], geneOrder: selected, heatmapData: [] };
-
-    // Get all hidden cells
+  const downsample = (groupBy, max = 1000) => {
+    // Find all hidden cells.
     const hiddenCellIds = union(Array.from(hidden), properties);
 
-    // Downsample cells.
-    let totalCellsToShow = 1000;
-    const groupByCluster = {};
+    // Get how many cells we are sampling.
+    let total = 0;
 
-    const louvainClusters = hierarchy.filter((clusters) => clusters.key === 'louvain');
-    if (louvainClusters.length > 0) {
-      let totalCellsFetched = 0;
+    // Create an object for storing the cells grouped by `groupBy`.
+    const groupedCells = {};
 
-      const clusterKeys = louvainClusters[0].children;
+    // Find the `groupBy` root node.
+    const rootNode = hierarchy.filter((clusters) => clusters.key === groupBy);
```

Comment:
what is this `groupBy` key? Is it a new addition to the redux store I have missed out from the previous PR?

Reply:
no, `groupBy` is the parameter to `downsample`, see the entire code. `groupBy` is what the sampling is based on, i.e. it will try to preserve a proportionally equal number of cells from that root node.